Fix #6 add initial api specification

softgitron opened the pull request:
**Short summary of changes in this pull request:**
This pull request provides main core and module API definitions in OpenAPI format. OpenAPI format is very useful, since it enables automatic checking of incoming API calls and provides a way to generate documentation of the API automatically.

**Short summary of important/debatable decisions and other notable choices made when working on the pull request:**
OpenAPI was selected as a language for defining APIs in potku-console.

**Other relevant information:**
This is very big and important feature. API can be partly altered in the future, but it will be increasingly harder when code base grows. Because of that all discussion is especially welcome related to this PR.

softgitron committed:
Add initial API specification

softgitron committed:
Openapi complete draft #1

softgitron commented:
First full core API draft is now available. Please review it. [OpenAPI editor](https://editor.swagger.io/) can be used for easier visual inspection of the specification.

softgitron commented:
> First full core API draft is now available. Please review it. [OpenAPI editor](https://editor.swagger.io/) can be used for easier visual inspection of the specification.

Initial API specification for service modules has been made as well.

softgitron committed:
Openapi for service modules, draft 1

EddieTheCubeHead commented on src/modules ipc library/OpenAPI/openapi_service_modules.yaml:
```diff
@@ -0,0 +1,204 @@
+    get:
+      tags:
+      - poll queries
+      summary: "Check that module is still active. Called by core on a regular intervall."
```
interval.* (Remove second l)

EddieTheCubeHead commented on src/modules ipc library/OpenAPI/openapi_service_modules.yaml:
```diff
@@ -0,0 +1,204 @@
+    post:
+      tags:
+      - poll queries
+      summary: "Get current active user count on the service."
```
of* the service

EddieTheCubeHead commented on src/core/OpenAPI/openapi.yaml:
```diff
@@ -0,0 +1,1793 @@
+- name: "services"
+  description: "Endpoint for controlling services (game, HTTP... servers)."
+- name: "servers"
+  description: "Endpoint for controlling virtual maschines, containers and bare metal maschines that actually run services."
```
virtual machines* ... ... bare metal machines

EddieTheCubeHead reviewed (commented):
Minor typos otherwise really solid.

softgitron committed:
Specification improvements and 1. cloud module API

softgitron committed:
1st complete cloud module API draft

softgitron committed:
Fix design flaws found during review session

softgitron committed:
Move definitions under docs for future online man

softgitron committed:
Second draft for all APIs
- Added bound checking
- Added format specifiers
- Added required markings
- Autoformated YAML files using Redhat formatter

softgitron commented:
Second draft of the APIs is now ready. Specification should be now pretty ready for prime time, but all comments are always welcome.

There might be still some rough edges in this specification, but I think that minor faults can be smoothed out during development process.

softgitron committed:
Add owners for the groups and permission alter ACL

softgitron committed:
Add token requirement for module APIs

softgitron committed:
Fix uuid data types

softgitron committed:
Change SSH keys to server keys for future needs

EddieTheCubeHead reviewed (dismissed):
LGTM

softgitron committed:
Add explicit action for post requests
- Can be used for recognizing, what action cloud module should take
